worlds/hk/__init__.py

```diff
@@ -441,12 +443,44 @@ def set_rules(self):
             world.completion_condition[player] = lambda state: state.count("Defeated_Pantheon_5", player)
         elif goal == Goal.option_godhome_flower:
             world.completion_condition[player] = lambda state: state.count("Godhome_Flower_Quest", player)
+        elif goal == Goal.option_grub_hunt:
+            pass  # will set in pre_fill()
         else:
             # Any goal
-            world.completion_condition[player] = lambda state: state._hk_can_beat_thk(player) or state._hk_can_beat_radiance(player)
+            world.completion_condition[player] = lambda state: state._hk_can_beat_thk(player) or state._hk_can_beat_radiance(player) or state.count("Defeated_Pantheon_5", player)
 
         set_rules(self)
 
+    def pre_fill(self):
+        grub_hunt_goal = self.options.GrubHuntGoal
+        goal = self.options.Goal
+        if goal in ["any", "grub_hunt"]:
+            def set_goal(grub_rule: typing.Callable[[CollectionState], bool]):
+                if goal == "grub_hunt":
+                    self.multiworld.completion_condition[self.player] = grub_rule
+                else:
+                    old_rule = self.multiworld.completion_condition[self.player]
+                    self.multiworld.completion_condition[self.player] = lambda state: old_rule(state) or grub_rule(state)
+
+            if grub_hunt_goal == grub_hunt_goal.special_range_names["all"]:
+                from collections import Counter
+                relevant_groups = self.multiworld.get_player_groups(self.player)
+                grub_player_count = Counter()
+
+                for grub in [item for item in self.multiworld.get_items() if item.name == "Grub"]:
+                    if grub.player in relevant_groups or grub.player == self.player:
+                        grub_player_count += Counter({grub.player: 1})
+                        if grub.location and grub.location.player in relevant_groups:
+                            # not counting our grubs stuck in item links because we also will count the group's copy
```

Comment:
you could just count them all at once across the multiworld, instead of hollow knight players times searching through the itempool by using stage_pre_fill

Reply:
i still need to loop through the grubs in the multiworld per player because their settings may be different, but moved it to stage_pre_fill so i can grab the grub list just once which did save time from what i could benchmark

Reply:
I'm actually a bit confused by this as well

It sounds like you could just

```
player_to_groups = {world.player: world.multiworld.get_player_groups(player) for world in worlds}
group_to_real_players = ... reverse the dict, don't remember how to off the top of my head ...
grubs_per_player = Counter()

for grub in grubs:
    player = grub.player
    if player in group_to_real_players:
        for real_player in group_to_real_players[player]:
            grubs_per_player[real_player] += 1
    else:
        grubs_per_player[player] += 1
```

And then you'd have a count of every player's grubs, all in one go. No?

Reply:
Actually I'm p sure I misunderstood that item links workaround... I'm not actually sure why working around item links is necessary. 
Either way though, I think I'd like a more detailed explanation as to why this doesn't work

Reply:
did take inspiration from your snippit and i think i have something like that working now but do not have the braincells to *test* to yet tonight as well,
but where is your group_to_real_players? because i got as far as 
```py
group_lookup = {}
for group_id, player in {(group_id, player) in multiworld.groups.items() for player in group["players"]}:
    if group_lookup[group_id]:
        group_lookup[group_id].add(player)
    else:
        group_lookup[group_id] = {player}
```
which i don't think is ideal?

Reply:
```
    if group_lookup[group_id]:
        group_lookup[group_id].add(player)
    else:
        group_lookup[group_id] = {player}
```

can be

```
group_lookup.setdefault(group_id, set()).add(player)
```

Or you could use a defaultdict.

Other than that, maybe there is some crazy comprehension magic that you could do? But I think I basically just do this in Witness

Reply:
Like, don't overthink it too much

```
group_lookup = defaultdict(set)
for group_id, players in multiworld.groups.items():
    for player in players:
        group_lookup[player].add(group_id)
```

Any comprehension you make probably just does this but in more complicated to understand syntax anyway